src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -54,6 +55,28 @@ public function createByTaxonPaginator(TaxonInterface $taxon, array $criteria =
     }
 
     /**
+     * Create paginator for products categorized under given product archetype.
+     *
+     * @param ArchetypeInterface $archetype
+     * @param array          $criteria
+     *
+     * @return \Pagerfanta\Pagerfanta
+     */
+    public function createByProductArchetypePaginator(ArchetypeInterface $archetype, array $criteria = array())
+    {
+        $queryBuilder = $this->getCollectionQueryBuilder();
+        $queryBuilder
+            ->innerJoin('product.archetype', 'archetype')
```

Comment:
you should use `->addSelect('archetype')` so that the query actually uses that join.